README.md
# FilterQL

[![Build Status](https://travis-ci.org/wearespindle/filterql.svg?branch=develop)](https://travis-ci.org/wearespindle/filterql)

With FilterQL you can write filter queries with the use of python objects.
These objects can be translated to json and send over the wire to be used
in some other backend. This allows you to not be limited to the way you
can define filters in (front-end) applications that communicate with
and API.

## Status

Currently actively used and watched.

## Usage

### Requirements

 * python 2.7
 * python 3.3, 3.4, 3.5
 * (optional) django >= 1.8

### Installation

```
pip install filterql
```

### Running

```python

from filterql import L, STARTSWITH

# Filter query for name to start with `spindle` or `devhouse`
lookup = L('name', 'spindle', lookup=STARTSWITH) | L('name', 'devhouse', lookup=STARTSWITH)

# Convert to json to be able to send it to other systems.
lookup_json = lookup.dumps()

# On the other side either convert back to L objects or to Django Q objects.

# L
l_filters = L.from_json(lookup_json)

# Django
from filterql.serializers import DjangoSerializer

django_filters = DjangoSerializer().from_json(lookup_json)
```

## Contributing

See the [CONTRIBUTING.md](https://github.com/wearespindle/filterql/blob/develop/CONTRIBUTING.md) file on how to contribute to this project.

## Contributors

See the [CONTRIBUTORS.md](https://github.com/wearespindle/filterql/blob/develop/CONTRIBUTORS.md) file for a list of contributors to the project.

## Roadmap

### Changelog

The changelog can be found in the [CHANGELOG.md](https://github.com/wearespindle/filterql/blob/develop/CHANGELOG.md) file.

### In progress

 * Publish on pypi

### Future

 * Improve logic for combining lookup elements
 * (De)serialization for other data types like XML


## Get in touch with a developer

If you want to report an issue see the [CONTRIBUTING.md](https://github.com/wearespindle/filterql/blob/develop/CONTRIBUTING.md) file for more info.

We will be happy to answer your other questions at opensource@wearespindle.com


CHANGELOG.md
# Changelog

## v0.2 (not released)

 * Fixed issue #1 with super self in exceptions

## v0.1

 * Initial release


tests/test_utils.py
from datetime import date, datetime

from pytest import raises

from filterql.exceptions import DecodeException
from filterql.lookup import TYPE_KEY, VALUE_KEY
from filterql.utils import (
    decode_date,
    decode_datetime,
    ENCODERS,
    TypeEncoder,
    type_decoder,
)


def test_type_encoder():
    """
    Test JSON encoding of Decimal, date and datetime.
    """
    date_obj = date(2017, 6, 6)
    the_input = {'date': date_obj}
    expected = '{"date": "%s"}' % date_obj.isoformat()

    result = TypeEncoder().encode(the_input)
    assert result == expected

    datetime_obj = datetime(2017, 6, 6, 13, 37, 0)
    the_input = {'datetime': datetime_obj}
    expected = '{"datetime": "%s"}' % datetime_obj.isoformat()

    result = TypeEncoder().encode(the_input)
    assert result == expected

    class Unknown():
        pass
    unknown_obj = Unknown()

    with raises(TypeError):
        TypeEncoder().encode(unknown_obj)


def test_type_decoder():
    """
    Test type decoding of date, datetime and Decimal.
    """
    _date = date(2017, 6, 6)
    date_dict = {VALUE_KEY: ENCODERS.get(type(_date))(_date), TYPE_KEY: date.__name__}
    decoded = type_decoder(date_dict)

    assert decoded.get(VALUE_KEY) == _date

    _datetime = datetime(2017, 6, 6, 13, 37, 0)
    datetime_dict = {VALUE_KEY: ENCODERS.get(type(_datetime))(_datetime), TYPE_KEY: datetime.__name__}
    decoded = type_decoder(datetime_dict)

    assert decoded.get(VALUE_KEY) == _datetime

    # Test decoding datetime with given type `date`.
    _datetime = datetime(2017, 6, 6)
    decimal_dict = {VALUE_KEY: ENCODERS.get(type(_datetime))(_datetime), TYPE_KEY: date.__name__}

    with raises(DecodeException):
        decoded = type_decoder(decimal_dict)


def test_decode_date():
    """
    Test decoding of date string.
    """
    date_obj = date(2017, 6, 6)

    result = decode_date(date_obj.isoformat())

    assert result == date_obj


def test_decode_datetime():
    """
    Test decoding of datetime string.
    """
    datetime_obj = datetime(2017, 6, 6, 13, 37, 0)

    result = decode_datetime(datetime_obj.isoformat())

    assert result == datetime_obj


tests/test_lookup.py
from datetime import date

from pytest import raises
import simplejson as json

from filterql.exceptions import InvalidFormat, InvalidValueException, UnsupportedLookupException
from filterql import ISNULL, L
from filterql.lookup import FIELD_KEY, LookupNode, LOOKUP_KEY, TYPE_KEY, VALUE_KEY
from filterql.lookup_types import LOOKUP_TYPES


def test_invalid_lookup_type():
    """
    Test a lookup with a invalid type.
    """
    message = '`spindle` is not a supported lookup! Supported lookups are: %s' % LOOKUP_TYPES

    with raises(UnsupportedLookupException) as excinfo:
        L('name', 'spindle', lookup='spindle')

    assert str(excinfo.value) == message


def test_invalid_value_for_lookup():
    """
    Test invalid value for a lookup.
    """
    message = 'Only True or False allowed for ISNULL lookup type.'

    with raises(InvalidValueException) as excinfo:
        L('name', 'null', lookup=ISNULL)

    assert str(excinfo.value) == message


def test_dumps():
    """
    Test dumping of a lookup.
    """
    # The JSON dumps is unordered so try all possible combinations.
    lookup = L('name', 'spindle')
    result = lookup.dumps()

    assert lookup.to_dict() == json.loads(result)
    assert lookup.to_dict() == L.from_json(result).to_dict()


def test_combining_invalid_types():
    """
    Test combining invalid types.
    """
    with raises(TypeError):
        L('name', 'spindle') & 1337


def test_lookup_node_kwarg_validation():
    """
    Test invalid types for LookupNode init.
    """
    with raises(TypeError):
        LookupNode(filters='x')

    with raises(TypeError):
        LookupNode(connector='hello')

    with raises(TypeError):
        LookupNode(negated='yes')


def test_one_lookup():
    """
    Test one normal lookup.
    """
    lookup = L('name', 'spindle')
    expected = {
        L.AND: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected


def test_one_not_lookup():
    """
    Test one `not` lookup.
    """
    lookup = ~L('name', 'spindle')
    expected = {
        L.NOT: {
            L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            ]
        }
    }

    assert lookup.to_dict() == expected


def test_and_lookup():
    """
    Test the `and` lookup.
    """
    lookup = L('name', 'spindle') & L('country', 'netherlands')
    expected = {
        L.AND: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected

    lookup = L('name', 'spindle') & L('country', 'netherlands') & L('status', 'awesome')
    expected = {
        L.AND: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected


def test_or_lookup():
    """
    Test the `or` lookup.
    """
    lookup = L('name', 'spindle') | L('name', 'devhouse')
    expected = {
        L.OR: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'name', VALUE_KEY: 'devhouse', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected

    lookup = L('name', 'spindle') | L('name', 'devhouse') | L('name', 'devhouse spindle')
    expected = {
        L.OR: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'name', VALUE_KEY: 'devhouse', LOOKUP_KEY: 'exact'},
            {FIELD_KEY: 'name', VALUE_KEY: 'devhouse spindle', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected


def test_one_not_and_lookup():
    """
    Test one normal and one `not` lookup.
    """
    lookup = L('name', 'spindle') & ~L('country', 'netherlands')
    expected = {
        L.AND: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            {L.NOT: {
                L.AND: [
                    {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
                ]
            }},
        ]
    }

    assert lookup.to_dict() == expected


def test_all_not_and_lookup():
    """
    Test all lookups with `not`.
    """
    lookup = ~(L('name', 'spindle') & L('country', 'netherlands'))
    expected = {
        L.NOT: {
            L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
            ]
        }
    }

    assert lookup.to_dict() == expected

    lookup = ~(L('name', 'spindle') & L('country', 'netherlands') & L('status', 'awesome'))
    expected = {
        L.NOT: {
            L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
            ]
        }
    }

    assert lookup.to_dict() == expected


def test_and_or_lookup():
    """
    Test lookup with the use of `and` and `or`.
    """
    lookup = L('name', 'spindle') & L('country', 'netherlands') | L('status', 'awesome')
    expected = {
        L.OR: [
            {L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
            ]},
            {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected

    lookup = L('status', 'awesome') | L('name', 'spindle') & L('country', 'netherlands')
    expected = {
        L.OR: [
            {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
            {L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
            ]},
        ]
    }

    assert lookup.to_dict() == expected


def test_complex_mix_lookup():
    """
    Test two complex mixed lookups with `not's` `and's` and `or's`.
    """
    lookup = (~L('name', 'spindle') & L('status', 'awesome')) & (L('country', 'netherlands') | L('country', 'germany'))
    expected = {
        L.AND: [
            {L.NOT: {L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
            ]}},
            {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
            {L.OR: [
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'germany', LOOKUP_KEY: 'exact'},
            ]},
        ]
    }

    assert lookup.to_dict() == expected

    lookup = (L('name', 'spindle') & L('status', 'awesome')) | ~(L('country', 'netherlands') | L('country', 'germany'))
    expected = {
        L.OR: [
            {L.AND: [
                {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'status', VALUE_KEY: 'awesome', LOOKUP_KEY: 'exact'},
            ]},
            {L.NOT: {L.AND: [{L.OR: [
                {FIELD_KEY: 'country', VALUE_KEY: 'netherlands', LOOKUP_KEY: 'exact'},
                {FIELD_KEY: 'country', VALUE_KEY: 'germany', LOOKUP_KEY: 'exact'},
            ]}]}},
        ]
    }

    assert lookup.to_dict() == expected


def test_duplicate_lookup():
    """
    Test when combining exactly the same lookups.
    """
    lookup = L('name', 'spindle') & L('name', 'spindle')
    expected = {
        L.AND: [
            {FIELD_KEY: 'name', VALUE_KEY: 'spindle', LOOKUP_KEY: 'exact'},
        ]
    }

    assert lookup.to_dict() == expected

    lookup = L('name', 'spindle') & L('name', 'spindle') & L('name', 'spindle')

    assert lookup.to_dict() == expected


def test_invalid_formats():
    """
    Test supplying invalid formats to from_dict.
    """
    wrong_format = {L.AND: '', L.OR: ''}
    message = 'Lookup root can only have 1 key %s' % wrong_format

    with raises(InvalidFormat) as execinfo:
        L.from_dict(wrong_format)

    assert str(execinfo.value) == message

    wrong_format = {L.NOT: {L.AND: '', L.OR: ''}}
    message = 'Lookup root can only have 1 key %s' % {L.AND: '', L.OR: ''}

    with raises(InvalidFormat) as execinfo:
        L.from_dict(wrong_format)

    assert str(execinfo.value) == message

    wrong_format = {'unknown': ''}
    message = 'Lookup root connector must be `%s`, `%s` or `%s` %s' % (L.AND, L.NOT, L.OR, wrong_format)

    with raises(InvalidFormat) as execinfo:
        L.from_dict(wrong_format)

    assert str(execinfo.value) == message

    wrong = {L.AND: 'wrong', L.OR: 'format'}
    wrong_format = {L.AND: [wrong]}
    message = 'Not a lookup or node for filter %s' % wrong

    with raises(InvalidFormat) as execinfo:
        L.from_dict(wrong_format)

    assert str(execinfo.value) == message


def test_adding_type_to_filter():
    """
    Test if type is added to filter for unsupported json types.
    """
    lookup = L('date', date(2017, 6, 6))

    _filter = lookup.filters[0]
    _type = _filter.get(TYPE_KEY)

    assert _type
    assert _type == date.__name__


tests/test_validators.py
from pytest import raises

from filterql import DAY, HOUR, ISNULL, MINUTE, MONTH, SECOND, WEEK
from filterql.exceptions import InvalidValueException
from filterql.validators import VALIDATORS


def test_isnull_validator():
    """
    Test the isnull_validator.
    """
    validator = VALIDATORS[ISNULL]
    message = 'Only True or False allowed for ISNULL lookup type.'

    validator(True)
    validator(False)

    with raises(InvalidValueException) as excinfo:
        validator('null')

    assert str(excinfo.value) == message


def test_month_validator():
    validator = VALIDATORS[MONTH]
    message = 'Month needs to be 1 <= month <= 12'

    for number in range(1, 13):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(0)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(13)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


def test_week_validator():
    validator = VALIDATORS[WEEK]
    message = 'Week needs to be 1 <= week <= 54'

    for number in range(1, 55):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(0)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(55)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


def test_day_validator():
    validator = VALIDATORS[DAY]
    message = 'Day needs to be 1 <= day <= 31'

    for number in range(1, 32):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(0)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(32)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


def test_hour_validator():
    validator = VALIDATORS[HOUR]
    message = 'Hour needs to be 0 <= hour <= 23'

    for number in range(24):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(-1)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(25)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


def test_minute_validator():
    validator = VALIDATORS[MINUTE]
    message = 'Minute needs to be 0 <= minute <= 59'

    for number in range(60):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(-1)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(60)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


def test_second_validator():
    validator = VALIDATORS[SECOND]
    message = 'Second needs to be 0 <= second <= 60'

    for number in range(61):
        validator(number)

    with raises(InvalidValueException) as excinfo:
        validator(-1)

    assert str(excinfo.value) == message

    with raises(InvalidValueException) as excinfo:
        validator(61)

    assert str(excinfo.value) == message

    message = 'Value needs to be an integer'
    with raises(InvalidValueException) as excinfo:
        validator('spindle')

    assert str(excinfo.value) == message


filterql/utils.py
from datetime import date, datetime
import traceback

from dateutil import parser
import simplejson as json

from .exceptions import DecodeException


class TypeEncoder(json.JSONEncoder):
    """
    Custome encoder for unsupported types.
    """
    def default(self, obj):
        """
        Encode specific types.

        Only executed on unknown JSON types.

        Args:
            obj: The obj of a unknown JSON type.
        """
        encode = ENCODERS.get(type(obj))

        if encode:
            return encode(obj)
        # Let the base class default method raise the TypeError
        return json.JSONEncoder.default(self, obj)


def type_decoder(dct):
    """
    Decodes a value if there is a type given for the value.

    Args:
        dct: The dict to decode.

    Returns:
        dict: The dict with certain types decoded.

    Raises:
        DecodeException: When decoding of the type failed.
    """
    # Avoid circular import.
    from .lookup import TYPE_KEY, VALUE_KEY

    value_type = dct.get(TYPE_KEY)
    if value_type:
        decode = DECODERS.get(value_type)
        if decode:
            try:
                dct[VALUE_KEY] = decode(dct[VALUE_KEY])
            except:
                trace = traceback.format_exc()
                raise DecodeException(
                    'Error decoding type `%s` %s\n%s' % (value_type, dct[VALUE_KEY], trace))
    return dct


def decode_date(date_string):
    """
    Decode a date string into a python date.

    Args:
        date_string (string): The string containing the date.

    Returns:
        date: The decoded date.
    """
    return datetime.strptime(date_string, '%Y-%m-%d').date()


def decode_datetime(datetime_string):
    """
    Decode a datetime string into a python datetime.

    Args:
        datetime_string (string): The string containing the datetime.

    Returns:
        datetime: The decoded datetime.
    """
    return parser.parse(datetime_string, fuzzy=False)


DECODERS = {
    date.__name__: decode_date,
    datetime.__name__: decode_datetime,
}

ENCODERS = {
    date: lambda x: x.isoformat(),
    datetime: lambda x: x.isoformat(),
}


filterql/lookup.py
import simplejson as json

from .exceptions import InvalidFormat, UnsupportedLookupException
from .lookup_types import EXACT, LOOKUP_TYPES
from .utils import ENCODERS, type_decoder, TypeEncoder
from .validators import VALIDATORS


KEY_PREFIX = '_'
FIELD_KEY = '%sfield' % KEY_PREFIX
LOOKUP_KEY = '%slookup' % KEY_PREFIX
VALUE_KEY = '%svalue' % KEY_PREFIX
TYPE_KEY = '%stype' % KEY_PREFIX


class LookupNode(object):
    """
    Node used for lookups. Has filters that can be either a dict with
    the given filter or another lookup node that contains more filters.
    """
    AND = '%sand' % KEY_PREFIX
    OR = '%sor' % KEY_PREFIX
    NOT = '%snot' % KEY_PREFIX
    default = AND

    def __init__(self, filters=None, connector=None, negated=False):
        """
        Constructs a new Node. With AND as default connector.
        """
        if not isinstance(filters, list) and filters is not None:
            raise TypeError('filters should be list or none')

        if connector not in (self.AND, self.OR) and connector is not None:
            raise TypeError('connector can only be `%s`, `%s` or `None`' % self.AND, self.OR)

        if not isinstance(negated, bool):
            raise TypeError('negated can only be of type bool')

        self.filters = filters[:] if filters else []
        self.connector = connector or self.default
        self.negated = negated

    def to_dict(self):
        """
        Function to transform the node and children into a dict recursively.

        Returns:
            dict: Dict of nodes and filters.
        """
        tree_dict = {}
        tree_dict[self.connector] = [f if isinstance(f, dict) else f.to_dict() for f in self.filters]

        if self.negated:
            return {self.NOT: tree_dict}

        return tree_dict

    def dumps(self):
        """
        Dump this node and children to json.

        Returns:
            string: The json string representing the lookup.
        """
        return json.dumps(self.to_dict(), cls=TypeEncoder, use_decimal=True)

    @staticmethod
    def from_json(l_json):
        """
        Function to create an instance from json.

        Args:
            l_json (string): The json string representing a lookup.

        Returns:
            LookupNode: The lookup instance based on the dict.
        """
        return L.from_dict(json.loads(l_json, object_hook=type_decoder, use_decimal=True))

    @staticmethod
    def from_dict(l_dict):
        """
        Function to create an instance from a dict.

        Args:
            l_dict (dict): The dict that represents a lookup.

        Returns:
            LookupNode: The lookup instance based on the dict.

        Raises:
            InvalidFormat: When the dict is not of the Lookup format.
        """
        # Set some defaults.
        negated = False
        connector = L.AND
        children = []

        # Check for valid formatting and get root connector key.
        keys = list(l_dict)
        if len(keys) != 1:
            raise InvalidFormat('Lookup root can only have 1 key %s' % l_dict)
        connector = keys[0]

        # Check if we have a `not` connector.
        if connector == L.NOT:
            negated = True
            # Get the `and` or `or` connector for the `not` connector.
            l_dict = l_dict.get(L.NOT)
            keys = list(l_dict)
            if len(keys) != 1:
                raise InvalidFormat('Lookup root can only have 1 key %s' % l_dict)
            connector = keys[0]

        # Check for (remaining) valid connector types.
        if connector not in [L.OR, L.AND]:
            raise InvalidFormat(
                'Lookup root connector must be `%s`, `%s` or `%s` %s' % (L.AND, L.NOT, L.OR, l_dict))

        filters = l_dict.get(connector)

        # Loop all filters in the lookup node.
        for _filter in filters:
            # Check if we have a filter or a nested lookup.
            valid_filter_keys = [FIELD_KEY, LOOKUP_KEY, VALUE_KEY]
            valid_type_filter_keys = valid_filter_keys + [TYPE_KEY]
            filter_keys = set(list(_filter))

            if len(filter_keys) == 1 and next(iter(filter_keys)) in (L.AND, L.OR, L.NOT):
                children.append(L.from_dict(_filter))
            elif (filter_keys == set(valid_filter_keys) or
                    filter_keys == set(valid_type_filter_keys)):
                children.append(_filter)
            else:
                raise InvalidFormat('Not a lookup or node for filter %s' % _filter)

        lookup = LookupNode()
        lookup.filters = children
        lookup.connector = connector
        lookup.negated = negated

        return lookup

    def __len__(self):
        """
        The number of filters in the node.
        """
        return len(self.filters)

    def add(self, other_l, conn_type):
        """
        Function to add a node to this node. Try to squash under certain
        conditions or just append them.

        Args:
            other_l (Lookup|LookupNode): The leaf/node to add to self..
            conn_type (string): The connector used in the adding.

        Returns:
            (Lookup|LookupNode): The result of the adding or squashing.
        """
        # Try to avoid double filters and combine nothing.
        if isinstance(other_l, Lookup) and other_l.filters[0] in self.filters:
            return

        # Try to squash filters when we can. When len of the filters of other
        # is 1 it does not matter what the connector of the other_l is. The
        # other_l is of the type Lookup and thus the connector is still
        # the default and subject to change either way.
        if (not other_l.negated and (other_l.connector == conn_type or len(other_l) == 1)):
            self.filters.extend(other_l.filters)
        else:
            self.filters.append(other_l)

    def negate(self):
        """
        Negate the sense of the root connector.
        """
        self.negated = not self.negated

    def _combine(self, other, connector):
        """
        Combine two lookups into one node.

        Args:
            other (Lookup|LookupNode): The other object to combine with.
            connector (string): The connector used for the combining.

        Returns:
            LookupNode: The created node with the combined lookups.
        """
        if not isinstance(other, LookupNode):
            raise TypeError(other)

        obj = LookupNode(connector=connector)
        obj.add(self, connector)
        obj.add(other, connector)
        return obj

    def __or__(self, other):
        return self._combine(other, self.OR)

    def __and__(self, other):
        return self._combine(other, self.AND)

    def __invert__(self):
        obj = type(self)()
        obj.add(self, self.AND)
        obj.negate()
        return obj


class Lookup(LookupNode):
    """
    Class that creates the needed filters and is used as `leaf` for
    a node.
    """
    def __init__(self, field, value, lookup=EXACT, validate=True):
        """
        Validate the given args and init the lookup.
        """
        if validate:
            self._validate_lookup(lookup)
            self._validate_value(value, lookup)

        self.value = value
        self.field = field
        self.lookup = lookup

        total_filter = {
            FIELD_KEY: self.field,
            LOOKUP_KEY: self.lookup,
            VALUE_KEY: self.value,
        }

        if type(value) in list(ENCODERS):
            total_filter[TYPE_KEY] = type(value).__name__

        super(type(self), self).__init__(filters=[total_filter])

    def _validate_lookup(self, lookup):
        """
        Validate whether a supported lookup type is provided.

        Args:
            lookup (string): The lookup type to check.

        Raises:
            UnsupportedLookupException
        """
        if lookup not in LOOKUP_TYPES:
            raise UnsupportedLookupException(lookup)

    def _validate_value(self, value, lookup):
        """
        Validate value based on lookup type.

        Args:
            value (type): The value to validate.
            lookup (string): One of the LOOKUP_TYPES to validate the value for.

        Raises:
            InvalidValueException
        """
        validate = VALIDATORS.get(lookup)
        if validate:
            validate(value)

    def __invert__(self):
        obj = type(self)(self.field, self.value, self.lookup)
        obj.negate()
        return obj


L = Lookup


filterql/serializers.py
from .lookup import FIELD_KEY, L, LOOKUP_KEY, LookupNode, VALUE_KEY
from .lookup_types import EXACT


class DjangoSerializer():
    """
    Class for (de)serializing L dict or L json to django Q object.
    """
    CONNECTOR_MAP = {
        L.AND: 'AND',
        L.OR: 'OR',
    }

    def __init__(self):
        """
        Lazy set the DJANGO_SUFFIX_DELIMITER based on a Django constant.
        """
        from django.db.models.constants import LOOKUP_SEP
        self.DJANGO_SUFFIX_DELIMITER = LOOKUP_SEP

    def from_json(self, json_string):
        """
        Load a json string into a django Q object.

        Args:
            json_string (string): A valid json string.

        Returns:
            Q: Django Q filter.
        """
        return self.deserialize(L.from_json(json_string))

    def deserialize(self, l_object):
        """
        Deserialize a dict into django Q object.

        Args:
            l_object (LookupNode): The L object to deserialize.

        Returns:
            Q: Django Q filter.
        """
        # Lazy import to avoid conflicts when not using this django serializer.
        from django.db.models import Q

        # Set some defaults.
        negated = l_object.negated
        connector = self.CONNECTOR_MAP[l_object.connector]
        children = []

        filters = l_object.filters

        # Loop all filters in the lookup node.
        for _filter in filters:
            if isinstance(_filter, LookupNode):
                children.append(self.deserialize(_filter))
            else:
                children.append(self._convert_filter(_filter))

        query = Q()
        query.children = children
        query.connector = connector
        query.negated = negated

        return query

    def _convert_filter(self, filter_dict):
        """
        Function to convert the L format of a filter to the django Q format
        of a filter.

        Args:
            filter_dict (dict): Dict with the filter keys and values.

        Returns:
            tuple: With the lookup and the value.
        """
        field = filter_dict[FIELD_KEY]
        value = filter_dict[VALUE_KEY]
        lookup = filter_dict[LOOKUP_KEY]
        if lookup == EXACT:
            lookup = ''
        else:
            lookup = '%s%s' % (self.DJANGO_SUFFIX_DELIMITER, lookup)

        django_lookup = '%s%s' % (field, lookup)

        return (django_lookup, value)


filterql/validators.py
from .lookup_types import DAY, HOUR, ISNULL, MINUTE, MONTH, SECOND, WEEK
from .exceptions import InvalidValueException


def integer_validator(value):
    if not isinstance(value, int):
        raise InvalidValueException('Value needs to be an integer')


def day_validator(day):
    integer_validator(day)

    if not (1 <= day <= 31):
        raise InvalidValueException('Day needs to be 1 <= day <= 31')


def month_validator(month):
    integer_validator(month)

    if not (1 <= month <= 12):
        raise InvalidValueException('Month needs to be 1 <= month <= 12')


def week_validator(week):
    integer_validator(week)

    if not (1 <= week <= 54):  # 54 happens every 28 years, next occurence 2028
        raise InvalidValueException('Week needs to be 1 <= week <= 54')


def hour_validator(hour):
    integer_validator(hour)

    if not (0 <= hour <= 23):
        raise InvalidValueException('Hour needs to be 0 <= hour <= 23')


def minute_validator(minute):
    integer_validator(minute)

    if not (0 <= minute <= 59):
        raise InvalidValueException('Minute needs to be 0 <= minute <= 59')


def second_validator(second):
    integer_validator(second)

    if not (0 <= second <= 60):  # 60 to account for leap seconds
        raise InvalidValueException('Second needs to be 0 <= second <= 60')


def isnull_validator(value):
    if not isinstance(value, bool):
        raise InvalidValueException(
            'Only True or False allowed for ISNULL lookup type.')


VALIDATORS = {
    # Date/time related validators.
    MONTH: month_validator,
    WEEK: week_validator,
    DAY: day_validator,
    HOUR: hour_validator,
    MINUTE: minute_validator,
    SECOND: second_validator,

    # Miscellaneous validators.
    ISNULL: isnull_validator,
}
